src/org/sasylf/refactor/RenameTheoremWizard.java

```diff
@@ -0,0 +1,229 @@
+	private RefactoringContext context;
+	private Map<IDocument, TextEdit> changes;
+
+	public RenameTheoremWizard(RefactoringContext context) {
+		super();
+		setNeedsProgressMonitor(true);
+		this.context = context;
+		changes = new HashMap<IDocument, TextEdit>();
+	}
+
+	@Override
+	public void addPages() {
+		page = new RenameTheoremWizardPage(context);
+		addPage(page);
+	}
+
+	/**
+	 * This method is called when 'Finish' button is pressed in
+	 * the wizard. We will create an operation and run it
+	 * using wizard as execution context.
+	 */
+	@Override
+	public boolean performFinish() {
+		try {
+			getOriginalFileChange();
+			collectReferencedChanges();
+			makeChanges();
+			PlatformUI.getWorkbench().saveAllEditors(false);	// save all changed editors
+			return true;
+		} catch (Exception e) {
+			e.printStackTrace();
+			return false;
+		}
+	}
+
+	private void getOriginalFileChange() throws BadLocationException {
+		IDocument doc = context.getProofDocument();
+		MultiTextEdit edit = new MultiTextEdit();
+		IRegion moduleLoc = null;
+		String newName = page.getTheoremName();
+
+		Map<String, ? super RuleLike> map = new HashMap<>();
+		context.getCompUnit().collectRuleLike(map);
+		Theorem theorem = (Theorem)map.get(context.getOldName());
+
+		moduleLoc = getModuleLocation(doc, theorem);
+
+		int offset = moduleLoc.getOffset();
+		int length = moduleLoc.getLength();
+
+		if (doc.get(offset, length).equals(newName)) {
+			// no change needed
+			return;
+		}
+		System.out.println("Replacing " + doc.get(offset, length) + " with " + newName);
+		edit.addChild(new ReplaceEdit(offset, length, newName));
+
+		List<QualName> qualNames = new ArrayList<>();
+		Consumer<QualName> consumer = name -> {			
+			if (name.getLastSegment().equals(context.getOldName())) {
+				QualName source = name.getSource();
+				if (source == null) {
+					qualNames.add(name);
+				}
+			}
+		};
+
+		context.collectQualNames(consumer);
+
+		for (int i = qualNames.size() - 1; i >= 0; --i) {
+			QualName name = qualNames.get(i);
+
+			moduleLoc = getModuleLocation(doc, name);
+
+			offset = moduleLoc.getOffset();
+			length = moduleLoc.getLength();
+
+			if (doc.get(offset, length).equals(newName)) {
+				// no change needed
+				continue;
+			}
+			System.out.println("Replacing " + doc.get(offset, length) + " with " + newName);
+			edit.addChild(new ReplaceEdit(offset, length, newName));
+		}
+		changes.put(doc, edit);
+	}
+
+	// get the change in the specified file
+	private void getFileChange(IFile file, IDocument doc) throws BadLocationException {
+		CompUnit cu = IDEProof.getCompUnit(file);
+		if (cu == null) return;
+		MultiTextEdit edit = new MultiTextEdit();
+		IRegion moduleLoc = null;
+		String newName = page.getTheoremName();
+		List<QualName> qualNames = new ArrayList<>();
+		Consumer<QualName> consumer = name -> {
+			if (name.getLastSegment().equals(context.getOldName())) {
+				QualName source = name.getSource();
+				if (source != null) {
```

Comment:
Could you instead resolve name to see if it was the theorem indicated?  Then this and the previous could perhaps be merged?